frontend/dockerfile/dockerfile_mount_test.go

```diff
@@ -221,3 +225,126 @@ RUN --mount=type=cache,target=/mycache [ -f /mycache/foo ]
 	}, nil)
 	require.NoError(t, err)
 }
+
+func testMountEnvVar(t *testing.T, sb integration.Sandbox) {
+	f := getFrontend(t, sb)
+
+	dockerfile := []byte(`
+FROM busybox
+ENV SOME_PATH=/mycache
+RUN --mount=type=cache,target=/mycache touch /mycache/foo
+RUN --mount=type=cache,target=$SOME_PATH [ -f $SOME_PATH/foo ]
+`)
+
+	dir, err := tmpdir(
+		fstest.CreateFile("Dockerfile", dockerfile, 0600),
+	)
+	require.NoError(t, err)
+	defer os.RemoveAll(dir)
+
+	c, err := client.New(context.TODO(), sb.Address())
+	require.NoError(t, err)
+	defer c.Close()
+
+	_, err = f.Solve(context.TODO(), c, client.SolveOpt{
+		LocalDirs: map[string]string{
+			builder.DefaultLocalNameDockerfile: dir,
+			builder.DefaultLocalNameContext:    dir,
+		},
+	}, nil)
+	require.NoError(t, err)
+}
+
+func testMountArg(t *testing.T, sb integration.Sandbox) {
+	f := getFrontend(t, sb)
+
+	dockerfile := []byte(`
+FROM busybox
+ARG MNT_TYPE=cache
+RUN --mount=type=$MNT_TYPE,target=/mycache2 touch /mycache2/foo
+RUN --mount=type=cache,target=/mycache2 [ -f /mycache2/foo ]
+`)
+
+	dir, err := tmpdir(
+		fstest.CreateFile("Dockerfile", dockerfile, 0600),
+	)
+	require.NoError(t, err)
+	defer os.RemoveAll(dir)
+
+	c, err := client.New(context.TODO(), sb.Address())
+	require.NoError(t, err)
+	defer c.Close()
+
+	_, err = f.Solve(context.TODO(), c, client.SolveOpt{
+		LocalDirs: map[string]string{
+			builder.DefaultLocalNameDockerfile: dir,
+			builder.DefaultLocalNameContext:    dir,
+		},
+	}, nil)
+	require.NoError(t, err)
+}
+
+func testMountEnvAcrossStages(t *testing.T, sb integration.Sandbox) {
+	f := getFrontend(t, sb)
+
+	dockerfile := []byte(`
+FROM busybox as stage1
+
+ENV MNT_ID=mycache
+ENV MNT_TYPE2=cache
+RUN --mount=type=$MNT_TYPE2,id=$MNT_ID,target=/abcabc touch /abcabc/foo
+RUN --mount=type=$MNT_TYPE2,id=$MNT_ID,target=/cbacba [ -f /cbacba/foo ]
```

Comment:
This test should check different values. Currently for example if expansion doesn't work and always returns the same constant value, this test would still pass.

Reply:
Thanks, good catch